server/storage/schema/auth.go

```diff
@@ -111,6 +111,7 @@ func (atx *authBatchTx) Lock() {
 
 func (atx *authBatchTx) Unlock() {
 	atx.tx.Unlock()
+	atx.tx.Commit()
```

Comment:
It looks like a big change although it's just one line code change.

Reply:
Yeah, I added a comment for this line. Anyway, I revived `UnsafeGetAllUsers()`, which uses batchTx. So technically it's not needed anymore (and is safe in the previous versions). But keeping it would be good for defensive purpose so I want to keep the `Commit()` call. How do you think?

Reply:
It took me about 2 hour to figure out the root cause why `ReadTx can miss some writes issued by its predecessor BatchTx.`.

When users delete something, i.e. a user, from the [Auth](https://github.com/etcd-io/etcd/blob/f94af77d4ad17effb9208808ff82ba181f9e64cd/server/storage/schema/bucket.go#L52) bucket, actually etcd just deletes the data from boltDB, and it doesn't delete the data from the read buf at all. It's exactly the reason why the following readTx may still can read the data immediately after deleting it. Once the periodically commit goroutimes commits all pending data, then readTx can see the new data. 

But for the k/v data in the [Key](https://github.com/etcd-io/etcd/blob/f94af77d4ad17effb9208808ff82ba181f9e64cd/server/storage/schema/bucket.go#L43) bucket, there isn't such issue, because each time when users delete a k/v pair, etcd just add a new k/v pair with tombstone flag set. So it updates the writeBuf at [batch_tx.go#L364](https://github.com/etcd-io/etcd/blob/f94af77d4ad17effb9208808ff82ba181f9e64cd/server/storage/backend/batch_tx.go#L364), and the readBuf at [batch_tx.go#L310](https://github.com/etcd-io/etcd/blob/f94af77d4ad17effb9208808ff82ba181f9e64cd/server/storage/backend/batch_tx.go#L310). So my previous comment [issuecomment-1169362104](https://github.com/etcd-io/etcd/pull/13954#issuecomment-1169362104) is still valid.

Actually all other buckets have such issue as well, but if we use batchTx to read data, then we will not see such issue, such as lease. 

Anyway, I am OK to keep the `atx.tx.Commit()` for now.

Let me take care of the rest of stuff later in separate PR if needed.

Reply:
Yeah, I feel the semantics of ReadTx and BatchTx is sometime counterintuitive and needs careful coding. If you can check other usages, it's really great. I personally want to revisit the value of read/write buffer of transactions.

Reply:
> I personally want to revisit the value of read/write buffer of transactions.

Sure.